crates/ruff/src/rules/ruff/rules/implicit_optional.rs

```diff
@@ -145,28 +144,28 @@ enum TypingTarget<'a> {
     None,
     Any,
     Object,
-    ForwardReference,
     Optional,
+    ForwardReference(Expr),
     Union(Vec<&'a Expr>),
     Literal(Vec<&'a Expr>),
     Annotated(&'a Expr),
 }
 
 impl<'a> TypingTarget<'a> {
-    fn try_from_expr(expr: &'a Expr, semantic: &SemanticModel) -> Option<Self> {
+    fn try_from_expr(expr: &'a Expr, checker: &Checker) -> Option<Self> {
```

Comment:
Maybe pass `semantic` and `locator` directly, to avoid an additional dependency on `Checker`?